File: src/25_reverse_nodes_in_k_group/main.py

# Definition for singly-linked list.
class ListNode:
    # http://c.biancheng.net/view/2291.html 
    # 以下操作限制ListNode实例只能动态添加val和next成员或者val()和next()的方法
    __slots__ = ('val','next')
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    # Runtime: 44 ms, faster than 91.39% 
    # Memory Usage: 15.2 MB, less than 37.43% 
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        def reverseRange(head: ListNode, tail: ListNode) -> ListNode:
            # 遍历链表一般需要pre,cur，而翻转链表则需要使用三指针，pre，cur和nxt
            pre, cur = None, head
            # [1,2,3]
            while cur != tail:
              nxt = cur.next # 因为cur.next更新为pre，为了保证循环继续，故把next先保存下来
              cur.next = pre
              pre = cur #右移pre为当前node，当前node右移为下一个node
              cur = nxt
            return pre #while跳出时cur == tail，pre是tail左边的那个node，是新的head

        nxt = head
        # 想了半天，还是得先遍历到目标节点，反正时间复杂度都是O(n)，近似认为是post-order遍历
        # [1,2,3], k = 2, nxt->1, cnt=0, nxt->2; cnt=1, nxt->3; cnt=2 stop 
        # isReverse = True
        for cnt in range(k):
          if not nxt:
            # isReverse = False
            return head #这里直接返回head岂不是很香
          else:
            nxt = nxt.next
        # 分配每个“节点”的任务，这里的一个“节点”对应ListNode里面的k个节点，一个group
        # 左分支处理：翻转[head,nxt)之间的node，完了之后再处理右侧剩下的节点
        newHead = reverseRange(head, nxt) #head if not isReverse else reverseRange(head, nxt)
        # 由于已经有了head，nxt，那么只需要让reverse返回新的head就好了
        
        # 右分支处理：对剩余部分继续调用reverseKGroup
        kidHead = self.reverseKGroup(nxt, k)

        # 中间节点处理：将最后的结果进行重组
        head.next = kidHead # 大意了，写成head.nxt = kidHead了...
        # head.nxt = kidHead 由于加入了__slots__ = ('val','next')，这样会让head.nxt报错
        return newHead

def printNode(root: ListNode):
  cur = root
  print("[",end="")
  while cur:
    print("%d," % cur.val, end="")
    cur = cur.next
  print("]")

if __name__ == "__main__":
  gua = Solution()
  n1 = ListNode(1)
  n2 = ListNode(2)
  n3 = ListNode(3)
  n4 = ListNode(4)
  n5 = ListNode(5)
  n1.next = n2; n2.next = n3
  n3.next = n4; n4.next = n5

  rtn = gua.reverseKGroup(n1, 2)
  printNode(rtn)

File: src/25_reverse_nodes_in_k_group/main.cpp
#include <iostream>

using namespace std;

/* Definition for singly-linked list. */
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    // 思路1: 大致还是divide & conquer思路,每次只reverse k这个部分的链表,preorder DFS
    // 思路2~: 先采用比较直观的递归调用方式,递归的话函数接口包括输入和返回
    //        输入可能是链表头,k值,k之后的链表尾部; 返回可能是链表整体的头
    // 思路3~: 由于是单向链表,为了获得链表尾部,得有一次循环操作; 思路3~需要与思路2~迭代
    // 思路4: 最终决定将遍历k个节点的操作放在递归函数里面, 这样输入就可以不需要考虑链表尾
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* tmp = helper(head, k);
        return tmp;
    }

    ListNode* helper(ListNode* head, int k) {
        ListNode *cur = head;
        // 把head到cur之间到这部分节点当作二叉树里面的中间节点, 然后现对它进行处理
        // 所以是pre order DFS
        // (1) 遍历k个节点, 如果不足k, 此时的head就是最终的head
        for (int i=0; i<k; i++) {
            if (!cur) return head;
            // 注意cur指向的是k长度链之外(特别强调)的那个node
            cur = cur->next;
        }

        // (2) 处理当前"节点", 即reverse k个链表node
        ListNode *new_head = reverse(head, cur);
        // 注意new_head是先前k长度链之中最后的node

        // (3) 处理叶子结点, 即链表剩下的部分
        ListNode *rst_head = reverseKGroup(cur, k);
        // 注意1: 要从k长度链之外的那个node开始
        // 注意2: 还要将rst_head添加翻转之后的new_head链表的尾部
        //       这里精妙绝伦的地方来了, 并没有再循环一次而是利用了head
        //       因为reverse之后, 原本head指向的节点已变为k个链表段的最后节点
        head->next = rst_head;

        return new_head;
    }

    // reverse的接口: 输入应该包括链表头和尾, 不清楚怎么搞先传这两个值进去
    // 输出应该是调整好之后的链表头; 注意原本的tail后面的连接不要丢失了
    ListNode* reverse(ListNode* head, ListNode* tail) {
        // 这里是已知链表头尾, 对链表进行的翻转; 有点loop的感觉
        ListNode* pre = tail;
        while (head != tail) {
            ListNode *tmp = head->next;
            head->next = pre;
            pre = head;
            head = tmp;
        }
        // 太经典了, 返回的是pre!!!!
        return pre;
    }
};

ListNode * generator(int N) {
    ListNode* rst = NULL;
    ListNode* pre = NULL;
    for (int i =0; N >=0 && i < N; i++) {
        ListNode* tmp = new ListNode(i+1);
        if (!rst) {
            rst = tmp;
            pre = tmp;
        } else {
            pre->next = tmp;
            pre = tmp;
        }
    }
    return rst;
}

int main() {
    int k;
    Solution gua;
    // case1
    ListNode* head1 = generator(k=5);
    ListNode* rst = gua.reverseKGroup(head1, k=2);
    ListNode* tmp = rst;
    while (tmp) {
        cout << tmp->val << "->";
        tmp = tmp->next;
    }
    cout << endl;

    cout << "finish!" << endl;
    return 0;
}